src/connection.js

```diff
@@ -599,6 +633,45 @@ class Connection extends EventEmitter {
       this.emit('charsetChange', token.newValue);
     });
 
+    this.tokenStreamParser.on('fedAuthInfo', (token) => {
+      const clientId = '7f98cb04-cd1e-40df-9140-3bf7e2cea4db';
+      if (token.fedAuthInfoData.stsurl && token.fedAuthInfoData.spn) {
+        this.fedAuthInfo.responsePending = true;
+        var context = new AuthenticationContext(token.fedAuthInfoData.stsurl);
+        context.acquireTokenWithUsernamePassword(token.fedAuthInfoData.spn, this.config.userName, this.config.password, clientId, (err, tokenResponse) => {
+          if (err) {
+            this.fedAuthInfo.responsePending = false;
+            //console.log('error');
+            this.loginError = ConnectionError('Security token could not be authenticated or authorized.', 'EFEDAUTH');
+          } else {
+            this.fedAuthInfo.responsePending = false;
+            this.fedAuthInfo.token = tokenResponse;
+          }
+        });
+      }
+    });
+
+    this.tokenStreamParser.on('featureExtAck', (token) => {
+      let value = undefined;
+      if ((value = token.featureAckOpts.get(FEDAUTH_OPTIONS.FEATURE_ID)) !== undefined) {
+        if (!this.fedAuthInfo.fedAuthInfoRequested) {
+          throw new Error('Did not request federated authentication, but received the acknowledgment');
+        }
+        switch (this.fedAuthInfo.fedAuthLibrary) {
+          case FEDAUTH_OPTIONS.LIBRARY_ADAL:
+            if (0 !== value.length) {
+              throw new Error(`Federated authentication acknowledgment for ${this.fedAuthInfo.method} authentication method includes extra data`);
+            }
+            break;
+          default:
+            throw new Error('Attempting to use unknown federated authentication library');
```

Comment:
How about making all these Errors a part of `ConnectionError` and emitting them in `connect` event?